xarray/core/datatree_io.py

```diff
@@ -1,7 +1,14 @@
+from __future__ import annotations
+
+from collections.abc import Mapping, MutableMapping
+from os import PathLike
+from typing import Any
+
 from xarray.core.datatree import DataTree
+from xarray.core.types import NetcdfWriteModes, ZarrWriteModes
 
 
-def _get_nc_dataset_class(engine):
+def _get_nc_dataset_class(engine: str | None):
```

Comment:
Shouldn't this be T_Engine ?
https://github.com/pydata/xarray/blob/4e9d557d47bf6792937792426559747204fce5ed/xarray/backends/api.py#L61C5-L67

Reply:
Good spotting - I definitely like using a `Literal` here rather than `str`. There does seem to be a slight mismatch between the possibly options for `T_NetcdfEngine` and the values that `_datatree_to_netcdf` validates against.

```
T_NetcdfEngine = ["netcdf4", "scipy", "h5netcdf"]
```

whereas the [check in this class](https://github.com/pydata/xarray/pull/9011/files#diff-0fe3a6e14024d8b4ba3bf84a137267fd53bc25475a004a8fcf02c01d1cfeb77aR53) is:

```
if engine not in [None, "netcdf4", "h5netcdf"]:
    raise ValueError("to_netcdf only supports the netcdf4 and h5netcdf engines")
```

I guess the question (again probably for @TomNicholas) - is the lack of "scipy" as an engine intentional? If so, I'll create a `Literal` type just in this file. If not, I'll migrate the `T_NetcdfEngine` to `xarray/core/types.py` and import it to use it here.

Reply:
`scipy` does not appear [to support groups](https://docs.scipy.org/doc/scipy/reference/generated/scipy.io.netcdf_file.html), so that may be intentional (or correct by accident, not sure)

Reply:
[6ed149](https://github.com/pydata/xarray/pull/9011/commits/6ed1491a12da1c6376040639a6502f13c27aeaba)

I'm not sure if I like the `get_args` implementation, so definitely open to better ideas there.

Reply:
It feels better than the alternative to me 
```
getattr(T_DataTreeNetcdfEngine, '__args__')

```

`get_args` is from the typing lib.  Really the only other option is to remove that as a type and go back to the literal strings?

Reply:
`get_args` fits well here, I think we do something similar in the interpolation code. 
The docs shows pretty much this use case: https://docs.python.org/3/library/typing.html#typing.get_args